src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -72,195 +72,248 @@ public static FieldType GetFieldDefinition(ISchema schema, IObjectGraphType pare
-                    {
-                        throw new InvalidVariableError(variableName, $"Unable to convert '{value.Value}' to '{type.Name}'", ex);
-                    }
-
-                    if (conversionFailed)
-                        throw new InvalidVariableError(variableName, $"Unable to convert '{value.Value}' to '{type.Name}'");
+                    ret = scalarGraphType.ParseValue(value);
                 }
-                else
+                catch (Exception ex)
                 {
-                    bool conversionFailed;
-
-                    try
-                    {
-                        conversionFailed = scalar.ParseValue(input) == null;
-                    }
-                    catch (Exception ex)
-                    {
-                        throw new InvalidVariableError(variableName, $"Unable to convert '{input}' to '{type.Name}'", ex);
-                    }
-
-                    if (conversionFailed)
-                        throw new InvalidVariableError(variableName, $"Unable to convert '{input}' to '{type.Name}'");
+                    throw new InvalidVariableError(variableName, $"Unable to convert '{value}' to '{scalarGraphType.Name}'", ex);
                 }
 
-                return;
+                if (ret == null)
+                    throw new InvalidVariableError(variableName, $"Unable to convert '{value}' to '{scalarGraphType.Name}'");
+
+                return ret;
             }
 
-            if (type is ListGraphType listType)
+            static object ParseValueList(ListGraphType listGraphType, string variableName, object value)
             {
-                var listItemType = listType.ResolvedType;
+                if (value == null)
+                    return null;
 
-                if (input is IEnumerable list && !(input is string))
+                // note: a list can have a single child element which will automatically be intrepreted as a list of 1 elements. (see below rule)
+                // so, to prevent a string as being intrepreted as a list of chars (which get converted to strings), we ignore considering a string as an IEnumerable
+                if (value is IEnumerable values && !(value is string))
                 {
-                    var index = -1;
-                    foreach (var item in list)
-                        AssertValidVariableValue(schema, listItemType, item, $"{variableName}[{++index}]", hasDefaultValue);
+                    // create a list containing the parsed elements in the input list
+                    var valueOutputs = new List<object>(values is ICollection collection ? collection.Count : 0);
+                    int index = 0;
+                    foreach (var val in values)
+                    {
+                        // parse/validate values as required by graph type
+                        valueOutputs.Add(ParseValue(listGraphType.ResolvedType, variableName + "[" + index++ + "]", val));
```

Comment:
Wow. What will happen on arrays of hundreds of elements? It seems to me that `variableName` is used only in error messages so let's pass additional `int?` argument in `ParseValue` like we did it for `BuildExecutionNode(ExecutionNode parent, IGraphType graphType, Field field, FieldType fieldDefinition, int? indexInParentNode = null)`

Reply:
Good idea. I’ll put this in a subsequent PR of optimizations.